### PR TITLE
Clang and clang tidy warning fixes

### PR DESCRIPTION
**Pull request by dimhotepus**
Small fixes based on warnings && suggestions from Clang && Clang-Tidy.

* Fix `signed` <> `unsigned` comparison mismatch (`-Wsign`).
* Mark single arg constructors as explicit where possible to disallow implicit conversions.
* Use `std::make_unique` instead of `reset(new X` to hide low-level memory management.
* Use defaulted default constructors and destructors to simplify code gen.
* Mark move ctor and assignment as `noexcept` for potential performance gain. See https://stackoverflow.com/questions/10787766/when-should-i-really-use-noexcept, "noexcept can dramatically improve performance of some operations..."
* Replace `std2::make_unique` with `std::make_unique` as C++14 already has it.
* Fix typo, remove unneeded `return` && headers, simplify `x == false` for `bool` x.
* Do not wrap queue size (`size_t`) around `UINT_MAX`.
* Add some tests.

### CODE REVIEW

**Comment by KjellKod** [comment]
Looks good.

**Comment by KjellKod** [comment]
Please incorporate the latest from master so we see that the CI tests works with these changes.

**Comment by dimhotepus** [comment]
@KjellKod Thanks for CI, finished with pr. Could you, please, verify all is good?

File: src/std2_type_traits.hpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * with no warranties. This code is yours to share, use and modify with no
+ * strings attached and no restrictions or obligations.
+ * 
+ * These type traits will be in C++17.
+ * 
+ * PUBLIC DOMAIN and NOT under copywrite protection.
+ *
+ * ********************************************* */
+
+#ifndef STD2_TYPE_TRAITS_HPP_
```
Comment (KjellKod) [inline src/std2_type_traits.hpp:12]: Can you replace this for #pragma once instead?

**Review by KjellKod — APPROVED** [review]
I'll remove the ifdef's. Great PR. Cool to see that someone is using this :)

File: src/std2_type_traits.hpp (hunk 1 above)
Comment (dimhotepus) [reply src/std2_type_traits.hpp:12]: You already done this. resolving :)

**Pull request merged**